expression/builtin_compare.go

```diff
@@ -1351,7 +1351,10 @@ func (c *compareFunctionClass) refineArgs(ctx sessionctx.Context, args []Express
 	isExceptional, finalArg0, finalArg1 := false, args[0], args[1]
 	isPositiveInfinite, isNegativeInfinite := false, false
 	// int non-constant [cmp] non-int constant
-	if arg0IsInt && !arg0IsCon && !arg1IsInt && arg1IsCon {
+    // Why check not null flag
+    // eg: int_col > const_val(which is less than min_int32)
+    // If int_col got null, compare result cannot be true
+	if arg0IsInt && !arg0IsCon && !arg1IsInt && arg1IsCon && mysql.HasNotNullFlag(arg0Type.Flag) {
 		arg1, isExceptional = RefineComparedConstant(ctx, *arg0Type, arg1, c.op)
 		finalArg1 = arg1
 		if isExceptional && arg1.GetType().EvalType() == types.ETInt {
```

Comment:
If `arg1 == false`, can we ignore the not-null flag ?

Reply:
select xxx where (col_int > (select sum(xx) xx) ) is null;

So need to distinguish null or false of `> expr`.